### ISSUE TITLE
prompt+completion training vs messages training for VLM

### ISSUE THREAD

**Issue by weathon** [huggingface/trl#4077]
### Reproduction

I have read this blog https://huggingface.co/learn/cookbook/fine_tuning_vlm_trl and tried to implement it myself. However, it seems like the model usually converges into a local solution. 
It uses this data format
```python
    return {
        "images": [sample["image"]],
        "messages": [
            {
                "role": "system",
                "content": [{"type": "text", "text": system_message}],
            },
            {
                "role": "user",
                "content": [
                    {
                        "type": "image",
                        "image": sample["image"],
                    },
                    {
                        "type": "text",
                        "text": sample["query"],
                    },
                ],
            },
            {
                "role": "assistant",
                "content": [{"type": "text", "text": sample["label"][0]}],
            },
        ],
    }
```
I then found (https://huggingface.co/docs/trl/en/sft_trainer#training-vision-language-models) and relized the prompt and completion should be in two different field to make it work. And I used this and it worked
```python
    images = [sample["image"].resize((512, 512))]
    prompt = [
        {
            "role": "user",
            "content": [
                {
                    "type": "image",
                    "image": sample["image"],
                },
                {
                    "type": "text",
                    "text": sample['prompt'],
                }
            ],
        },
    ]

    completion = [{
            "role": "assistant",
            "content": sample["completion"],
        }]
    return {
        "images": images,
        "prompt": prompt,
        "completion": completion,
    }
```
Is this intended or just version difference? I am training Qwen-2.5-VL-3B.

### System Info

- Platform: Linux-6.1.123+-x86_64-with-glibc2.35
- Python version: 3.12.11
- TRL version: 0.24.0.dev0
- PyTorch version: 2.8.0+cu126
- accelerator(s): NVIDIA A100-SXM4-40GB
- Transformers version: 4.56.1
- Accelerate version: 1.10.1
- Accelerate config: not found
- Datasets version: 4.0.0
- HF Hub version: 0.34.4
- bitsandbytes version: 0.47.0
- DeepSpeed version: not installed
- Diffusers version: 0.35.1
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.106.1
- PEFT version: 0.17.1
- vLLM version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
If you only use one column "messages", the loss is computed on the full sequence (not recommended for VLM because it also computes the loss on the image tokens.

The recommended way is to have one column for prompt, and one column for completion. In this case, the loss is computed on the completion part only.

**Comment by weathon** [comment]
Could we make this clear in the tutorial? It took me a long time to figure
out this.

**Comment by qgallouedec** [comment]
Thanks for reporting, and sorry for the inconvenience. To explain the context: when the blog was released, prompt-completion training wasn't supported.

Would you like to open a PR to fix it?

**Comment by weathon** [comment]
Sure I will start a PR on the cookbook repo

**Comment by albertvillanova** [comment]
Hey @weathon, sorry for the slow follow-up — catching up on the backlog. 

To close the loop: this is intended behavior, not a version bug. With a single `messages` column, SFT computes loss over the whole sequence (including image tokens for VLMs, since `assistant_only_loss` isn't supported for VLMs), whereas separate `prompt`/`completion` columns auto-enable completion-only loss so the prompt is masked out and loss is computed on the completion tokens only — see https://github.com/huggingface/trl/blob/fceadc75c81bd103a6cf7f870eafddf1106b6065/trl/trainer/sft_trainer.py#L1168-L1171 and https://github.com/huggingface/trl/blob/fceadc75c81bd103a6cf7f870eafddf1106b6065/trl/trainer/sft_trainer.py#L784-L785. That's why your second (prompt/completion) format trained well. The TRL docs you found already describe this; the outdated bit is the external cookbook blog, which you kindly offered to fix in a PR on huggingface/cookbook. 

Closing here since the TRL behavior is working as intended — thanks for the report, and feel free to re-open if anything else comes up!
